src/software/backend/replay_logging/replay_reader.cpp

```diff
@@ -0,0 +1,137 @@
+#include "replay_reader.h"
+
+#include <google/protobuf/util/delimited_message_util.h>
+
+#include <fstream>
+
+// unfortunately protobuf forces me to write caveman code
+extern "C"
+{
+#include <fcntl.h>
+#include <unistd.h>
+}
+
+namespace fs = std::experimental::filesystem;
+
+ReplayMsg ReplayReader::readDelimitedReplayProtobufFile(const fs::path& file_path)
+{
+    // imagine having to write caveman code to read protobuf msgs from files
+    ReplayMsg msg;
+    int fd = open(file_path.c_str(), O_RDONLY);
+
+    if (fd < 0)
+    {
+        throw std::invalid_argument(std::string("Failed to open() ") +
+                                    file_path.string() + " with error " +
+                                    std::to_string(fd));
+    }
+
+    auto file_input = std::make_unique<google::protobuf::io::FileInputStream>(fd);
+    auto coded_input =
+        std::make_unique<google::protobuf::io::CodedInputStream>(file_input.get());
+    bool result = google::protobuf::util::ParseDelimitedFromCodedStream(
```

Comment:
super nit: but can we define msg down here closer to where it's used? Makes it easier to read